### PR TITLE
plugin/kubernetes: Add in-cluster integration tests 

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. What does this pull request do?

This PR adds adds integration tests for kubernetes+autopath, and migrates most kubernetes tests to execute in-cluster.
Some background:  Currently during integration testing, coredns and test queries are all executed external to the kubernetes cluster. To correctly test kubernetes+autopath, we need an in-cluster deployment.  Specifically, coredns needs to be deployed in place of kube-dns in a kubernetes cluster, and queries need to be executed from within pods in the cluster.  This is a more typical/realistic deployment in general (not just for autopath tests), so most integration tests have been migrated to use this in-cluster deployment. The one exception is the out-of-cluster API failover test, which only applies to an out-of-cluster-deployemnts.

Minikube is used for the in-cluster deployment testing.  I investigated using hyperkube, and real kubernetes single node deployment, both had too many problems (hyperkube acting unstable, real k8s not supported in trusty).

Coredns is deployed in cluster by preparing the docker container and pushing to a local docker repo (which itself runs in a container).  The coredns deployment pulls from the local container repo.

In-cluster queries are made by spinning up a dns/tools pod, executing a query from the pod using `dig`, and then parsing the results into a dns.Msg struct, allowing us to re-use the existing test helpers.

I've split up tests into parallel jobs in travis, the longest taking about 5 minutes typically. 
The in-cluster jobs do a double compile, once for building the .  Figuring out how to avoid that could shave another minute off or so.

### 2. Which issues (if any) are related?

#902 

### 3. Which documentation changes (if any) need to be made?

None

### CODE REVIEW

File: .travis/kubernetes/coredns.yaml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  selector:
+    k8s-app: coredns
+  clusterIP: 10.0.0.10
+  ports:
+  - name: dns
+    port: 53
+    protocol: UDP
+  - name: dns-tcp
+    port: 53
+    protocol: TCP
```
Comment (johnbelamaric) [inline .travis/kubernetes/coredns.yaml:123]: Missing newline.

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
Looks like pulling the etcd tag out affected the etcd *unit* tests not just the integration tests.

**Comment by miekg** [comment]
I'm sorry: I don't want this merged in the main coredns repo for various reasons:

* double time spend in travis (3 -> 6 minutes)
* multiple timeout loops seen that can wait up to 5 minutes
* more external network dependencies (would like to rid of those, not add more)
* custom code that parses `dig` output and a lot more infrastructure code
* more dependencies on docker
* only works for 1 plugin: *kubernetes*

This should move to an already running cluster where we can pull in the code from a PR (webhooks) and then do the testing we need. For added fun we can do a chaos monkey testing there as well

**Comment by johnbelamaric** [comment]
To start why don't we create a test repo and move these integration tests to there? I think it's still easier to run minikube than have to have an existing running cluster, though we could move to that at some point if we think it makes the most sense.

I think we need to have this level of integration test that actually runs queries from inside a pod.

**Comment by miekg** [comment]
No disagreement about the need for these kind of tests. As said per
previous comment, other repo is fine and gives clearer ownership.

**Comment by chrisohaver** [comment]
@miekg, would you be OK with a solution if  ...
* it builds kubernetes env from scratch each time instead of relying on an already running one.
* coredns repo tests builds and pushes coredns docker image and pushes to docker-hub (in parallel with existing tests, non-failing if push fails), then the ci repo integration test pull the image and runs tests.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/kubern..." ]
>@miekg, would you be OK with a solution if  ...

I think we should be very careful adding code into the main coredns repo. If we 
play/polish this in another repo I care much less.

>* it builds kubernetes env from scratch each time instead of relying on an already running one.

What does that buy us?

>* coredns repo tests builds and pushes coredns docker image and pushes to docker-hub (in parallel with existing tests, non-failing if push fails), then the ci repo integration test pull the image and runs tests.

If we are doing it separately from travis, this is all fine. Might move the 
run-etcd checks into this eventually as well, and have travis only do the unit 
tests.

**Comment by johnbelamaric** [comment]
Building the K8s environment means we don't have to rely on something in the environment (an external cluster) to be set up and in a running state. If we use travis we have to set it up every time. If we use Packet we could leave one running, but then we may have issues doing parallel tests and we need to make sure everything is cleaned up before each run.

**Comment by chrisohaver** [comment]
Additionally, building the K8s environment on demand  allows us to test on different versions of k8s more easily.

**Pull request closed without merging**